A28-xds-traffic-splitting-and-routing.md

```diff
@@ -0,0 +1,1095 @@
+1.  For each new action, either find an action to reuse, or create a new action
+    1.  If there’s a to-be-removed action with the same set of clusters (but different weights), it will get the old action’s name, otherwise it will be assigned a new name.
+1.  To accelerate processing, use 2-level-maps to keep track of the actions
+    1.  First level key is cluster names without weight, second level is cluster names + weights
+    1.  See appendix for [pseudocode](#Action-name-reusing-pseudocode).
+
+
+#### Route list
+
+For each `{matchers, action}`, a route will be generated:
+
+*   With matchers in json representation.
+*   The action will be the name of this action from the action list.
+
+The processed response will be converted to a [JSON service
+config](#JSON-service-config-generated-by-xDS-resolver).
+
+### Multiple watches for the same resource
+
+There could be multiple instances of CDS LB policies, and EDS as a child of
+CDS. So the xDS client will get parallel requests for `WatchCluster()` and
+`WatchEndpoints()`. And those watches can be for the same or for different
+resource\_names. The xDS client should be able to handle them.
+
+#### Start a new watch
+
+When the client is asked to watch for a resource\_name:
+
+*   If there’s an existing watch for this resource\_name, the client will not
+    send a xDS request
+    *   If there’s already a response, the new watcher will be notified
+        immediately
+*   If this is a new resource\_name, the client will start the watch by sending a
+    xDS request
+    *   If there are existing watches for the same type, but different
+        resource\_names, the new xDS request’s resource\_names should be the the
+        new name appended to the old names
+    *   Note that this could cause the server to resend the data for the previous
+        resouce\_names. The xDS client should try to filter out the duplicate
+        responses, and not notify the watchers.
+
+#### Cancel a watch
+
+When a watch is cancelled:
+
+*   If there are still other watch for the same name, the client will do nothing
+*   If this is the last watch for this name
+    *   The client will send a xDS request, with this name removed from
+        resource\_names list
+        *   Note that this will result in a xDS request with an empty
+            resource\_names list when the last watch is cancelled. It's expected
+            that the server won't handle this as a wild card (because this is not
+            the first request).
+    *   The item for this resource\_name will be cleared from the cache
+        *   The cache should only contain the responses for the in-process
+            watches. If we keep the item in cache after the watch is cancelled,
+            it will quickly become outdated.
+
+#### Handle responses
+
+When a response is received, the xDS client will
+
+*   Parse and notify all the watchers for this type
+*   Store the updated in cache, to prepare for future watches for the same
```

Comment:
s/Store the updated in cache/Store the updated resource in the cache/

Reply:
Done